torchtune/data/_collate.py

```diff
@@ -12,7 +12,133 @@
 from torchtune.data._common import CROSS_ENTROPY_IGNORE_IDX
 
 
+def left_pad_sequence(
+    sequences: List[torch.Tensor],
+    batch_first: bool = True,
+    padding_value: float = 0,
+) -> torch.Tensor:
+    """
+    This function is identical to :func:`torch.nn.utils.rnn.pad_sequence`, but
+    instead pads a list of variable length Tensors from the left to the length
+    of the longest sequence.
+
+    Example:
+        >>> a = torch.tensor([1, 2, 3])
+        >>> b = torch.tensor([4, 5, 6, 7])
+        >>> c = torch.tensor([8, 9, 10, 11, 12])
+        >>> left_pad_sequence([a, b, c], padding_value=0)
+        tensor([[ 0,  0,  1,  2,  3],
+                [ 0,  4,  5,  6,  7],
+                [ 8,  9, 10, 11, 12]])
+
+    Note:
+        This function returns a Tensor of size ``T x B x *`` or ``B x T x *``
+        where `T` is the length of the longest sequence. This function assumes
+        trailing dimensions and type of all the Tensors in sequences are same.
+
+    Args:
+        sequences (List[torch.Tensor]): list of variable length sequences.
+        batch_first (bool, optional): if ``True``, the output will be in ``B x T x *``
+            format, ``T x B x *`` otherwise.
+        padding_value (float, optional): value for padded elements. Default: 0.
+
+    Returns:
+        Tensor of size ``T x B x *`` if :attr:`batch_first` is ``False``.
+        Tensor of size ``B x T x *`` otherwise
+    """
+    return pad_sequence(
+        map(lambda x: torch.flip(x, dims=[0]), sequences),
+        batch_first=batch_first,
+        padding_value=padding_value,
+    ).flip(dims=[1])
+
+
 def padded_collate(
+    batch: List[Dict[str, List[int]]],
+    *,
+    pad_fn: Callable,
+    keys_to_pad: List[str],
```

Comment:
I'm not 100% about how this should be paramterized and very open to suggestions. The current setup means all `keys_to_pad` will be collated with `pad_fn`, even if they don't need padding, which shouldn't have any effect when the key doesn't need padding e.g. in the case of classification tasks, but it's not super intuitive.

Reply:
After looking back on this, it looks silly. I've simplified.